### ISSUE TITLE
ambiguity of overloaded function call not reported

### ISSUE THREAD

**Issue by firewave** [llvm/llvm-project#118801]
```cpp
#include <cstddef>

class C;

static void func(long long);
static void func(const C*);


void f()
{
    func(std::size_t(0));
}
```

https://godbolt.org/z/T6G595Edd

The code above is accepted by Clang but fails to compile with GCC and MSVC.

```
<source>: In function 'void f()':
<source>:11:9: error: call of overloaded 'func(std::size_t)' is ambiguous
   11 |     func(std::size_t(0));
      |     ~~~~^~~~~~~~~~~~~~~~
<source>:5:13: note: candidate: 'void func(long long int)'
    5 | static void func(long long);
      |             ^~~~
<source>:6:13: note: candidate: 'void func(const C*)'
    6 | static void func(const C*);
      |             ^~~~
```

```
<source>(11): error C2668: 'func': ambiguous call to overloaded function
<source>(6): note: could be 'void func(const C *)'
<source>(5): note: or       'void func(__int64)'
<source>(11): note: while trying to match the argument list '(size_t)'
```

Clang 5 was the last version which reported this

```
<source>:11:5: error: call to 'func' is ambiguous
    func(std::size_t(0));
    ^~~~
<source>:5:13: note: candidate function
static void func(long long);
            ^
<source>:6:13: note: candidate function
static void func(const C*);
            ^
```

**Comment by MagentaTreehouse** [comment]
Looks like clang is correct: the `const C*` overload should not be viable, because `std::size_t(0)` is not an integer literal and cannot be converted to a pointer with a standard convertion, according to [[conv.ptr]/1](https://timsong-cpp.github.io/cppwp/conv.ptr#1)

**Comment by firewave** [comment]
Thanks for the analysis.

GCC has not implemented it yet: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=52145

Visual Studio requires `/permissive-`: https://developercommunity.visualstudio.com/t/Which-Zc-option-would-prevent-for-VARIA/10331612#T-N10334995
